src/ejabberd_local.erl

```diff
@@ -333,38 +320,27 @@ handle_cast(_Msg, State) ->
 %% Description: Handling all non call/cast messages
 %%--------------------------------------------------------------------
 handle_info({route, Acc, From, To, El}, State) ->
-    process_packet(Acc, From, To, El, undefined),
-    {noreply, State};
-handle_info({register_iq_handler, Host, XMLNS, Module, Function}, State) ->
-    ets:insert(?IQTABLE, {{XMLNS, Host}, Module, Function}),
-    catch mod_disco:register_feature(Host, XMLNS),
+    process_packet(Acc, From, To, El, #{}),
     {noreply, State};
-handle_info({register_iq_handler, Host, XMLNS, Module, Function, Opts}, State) ->
-    ets:insert(?IQTABLE, {{XMLNS, Host}, Module, Function, Opts}),
+handle_info({register_iq_handler, Host, XMLNS, IQHandler}, State) ->
+    ets:insert(?IQTABLE, {{XMLNS, Host}, IQHandler}),
     catch mod_disco:register_feature(Host, XMLNS),
     {noreply, State};
 handle_info({unregister_iq_handler, Host, XMLNS}, State) ->
     case ets:lookup(?IQTABLE, {XMLNS, Host}) of
-        [{_, Module, Function, Opts}] ->
-            gen_iq_handler:stop_iq_handler(Module, Function, Opts);
+        [{_, IQHandler}] ->
+            gen_iq_component:stop_iq_handler(IQHandler),
+            ets:delete(?IQTABLE, {XMLNS, Host}),
+            catch mod_disco:unregister_feature(Host, XMLNS);
```

Comment:
these catches would make life harder...
`try ... catch` with logging would be better.

Reply:
This is out of scope of this PR. mod_disco rework is planned for the future.